ClassFieldTheory/LocalCFT/NonArchDef.lean

```diff
@@ -0,0 +1,159 @@
+import Mathlib
+
+/-!
+# Definition of Non-Archimedean Local Fields
+
+We define non-archimedean local fields as a class `IsNonArchLF`.
+
+-/
+
+universe u v
+
+class IsNonArchLF (K : Type u) [Field K] [ValuativeRel K] [UniformSpace K] : Prop extends
+  IsTopologicalDivisionRing K,
+  IsUniformAddGroup K,
+  LocallyCompactSpace K,
+  CompleteSpace K,
+  ValuativeTopology K,
+  ValuativeRel.IsNontrivial K,
+  ValuativeRel.IsRankLeOne K,
+  ValuativeRel.IsDiscrete K
+
+open ValuativeRel
+
+namespace IsNonArchLF
+
+variable (p : ℕ) [Fact p.Prime]
+
+instance : LocallyCompactSpace ℚ_[p] := inferInstance
+
+instance : IsNonArchLF ℚ_[p] where
+  mem_nhds_iff := sorry
+  nonempty := sorry
+  has_maximal_element := sorry
+
+variable (K : Type u) [Field K] [ValuativeRel K] [UniformSpace K] [IsNonArchLF K]
+  (L : Type v) [Field L] [ValuativeRel L] [UniformSpace L] [IsNonArchLF L]
+
+instance : (Valued.v : Valuation K (ValueGroupWithZero K)).IsNontrivial :=
+  ValuativeRel.isNontrivial_iff_isNontrivial.mp inferInstance
+
+noncomputable
+instance : (Valued.v : Valuation K (ValueGroupWithZero K)).RankOne where
+  hom := IsRankLeOne.nonempty.some.emb
+  strictMono' := IsRankLeOne.nonempty.some.strictMono
+
+open scoped Valued in
+instance : ProperSpace K := ProperSpace.of_nontriviallyNormedField_of_weaklyLocallyCompactSpace K
+
+instance : IsDiscreteValuationRing 𝒪[K] :=
+  (Valued.integer.properSpace_iff_completeSpace_and_isDiscreteValuationRing_integer_and_finite_residueField.mp inferInstance).2.1
+
+variable {K} in
+def IsUniformiser (x : 𝒪[K]) : Prop :=
+  Irreducible x
+
+/-- An arbitrary choice of a uniformiser in a local field. -/
+noncomputable def ϖ : 𝒪[K] :=
+  (IsDiscreteValuationRing.exists_irreducible _).choose
+
+@[simp] lemma isUniformiser_ϖ : IsUniformiser (ϖ K) :=
+  (IsDiscreteValuationRing.exists_irreducible _).choose_spec
+
+lemma ϖ_ne_zero : ϖ K ≠ 0 :=
+  (IsDiscreteValuationRing.exists_irreducible _).choose_spec.ne_zero
+
+lemma associated_ϖ_iff_isUniformiser (x : 𝒪[K]) :
+    Associated x (ϖ K) ↔ IsUniformiser x :=
+  ⟨fun h ↦ h.symm.irreducible (isUniformiser_ϖ K),
+  fun h ↦ IsDiscreteValuationRing.associated_of_irreducible _ h (isUniformiser_ϖ K)⟩
```

Comment:
I don't think we should be making a global choice.

Reply:
ditto